src/models/generalized_mixture_reach_surface.py

```diff
@@ -0,0 +1,186 @@
+from typing import Iterable
+from wfa_planning_evaluation_framework.models.reach_point import ReachPoint
+from wfa_planning_evaluation_framework.models.reach_surface import ReachSurface
+from wfa_planning_evaluation_framework.models.reach_curve import ReachCurve
+
+solvers.options["show_progress"] = False
+
+
+class GeneralizedMixtureReachSurface(ReachSurface):
+    """Models reach with the generalized mixture overlap model.
+
+    GeneralizedMixtureReachSurface models the combined reach with the formula
+
+
+    N * sum_{j=1..c} [1 - prod_{i=1..p} (1-a_i_j r_i / N)] +epsilon
+
+    where :
+          all a_i_j >= 0
+          sum_{j=1..c} a_i_j=1 for all i.
+
+    """
+
+    def __init__(
+        self,
+        reach_curves: Iterable[ReachCurve],
+        reach_points: Iterable[ReachPoint],
+        num_clusters: int,
+        N: int,
+    ):
+        """Constructor for GeneralizedMixtureReachSurface.
+
+        Args:
+          reach_curves: A list of ReachCurves to be used in model fitting and
+            reach prediction.
+          reach_points: A list of ReachPoints to which the model is to be fit.
+            This list is of arbitrary length, but it should contain a minimum of
+            p(p-1)/2 points, where p is the number of publishers. Otherwise,
+            there is no guarantee of a unique solution. The reach points
+            represent arbitrary points on the reach surface, involving spend at
+            multiple advertisers
+          num_clusters: number of clusters. Different users in the same cluster
+            have the same probability to receive a new impression.
+          N : A non-negative value that is at least as large as the maximum
+            reach of any of the per-publisher reach functions,
+            e.g., N >= reach_curves[i].max_reach for i=1,2,...,p.
+        """
+
+        self._reach_curves = copy.deepcopy(reach_curves)
+        self._n = len(reach_points)
+        self._p = len(reach_points[0].impressions)
+        self._c = num_clusters
+        self._N = N
+        self._n_iter = 10
+        super().__init__(data=reach_points)
+
+    def update_theta_from_a(self, a, r):
+        """Udpate theta using equation:
+        theta(x_1, ... ,x_p) = (1 - prod_{i=1..p} (1- x_i)) / sum_{i=1..p} x_i}
+
+        Args:
+          a: a length (p * c) vector. Current best guess of a.
+          r: a length n list of length p vectors. Single publisher reaches for
+            each training point.
+
+        Returns:
+          A length n list of length c vectors that indicate each
+          theta_j^{(k)}.
+        """
+        theta = []
+        for k in range(self._n):
+            theta_k = np.zeros(self._c)
+            for j in range(self._c):
+                x = a[[j + i * self._c for i in range(self._p)]] * r[k] / self._N
+                theta_k[j] = (1 - np.prod(1 - x)) / np.sum(x)
+            theta.append(theta_k)
+        return theta
+
+    def solve_a_given_z(self, z, show_status=False):
```

Comment:
Inputs should be as follows:

y:  an n x 1 matrix of total reach values  
r:  an n x p matrix of per-publisher reach values
th: an n x c matrix of theta values
N: total universe size

Output:
a: a c x p matrix that minimizes minimizes ... subject to the constraints that a[i][j] >= 0 and row sums are one.